csharp/src/Microsoft.ML.OnnxRuntime/OrtTypeInfo.shared.cs

```diff
@@ -0,0 +1,272 @@
+                }
+                return _mapTypeInfo.Value;
+            }
+        }
+
+        /// <summary>
+        /// Fetches OptionalTypeInfo from the OrtTypeInfo iff this OrtTypeInfo represents a optional type.
+        /// </summary>
+        /// <returns>Instance of OptionalTypeInfo</returns>
+        /// <exception cref="OnnxRuntimeException"></exception>
+        public OrtSequenceOrOptionalTypeInfo OptionalTypeInfo
+        {
+            get
+            {
+                if (OnnxType != OnnxValueType.ONNX_TYPE_OPTIONAL)
+                {
+                    throw new OnnxRuntimeException(ErrorCode.InvalidArgument, "TypeInfo does not represent a optional");
+                }
+                return _sequenceOrOptional.Value;
+            }
+        }
+    }
+
+    /// <summary>
+    /// This struct represents type and shape information for a tensor.
+    /// It may describe a tensor type that is a model input or output or
+    /// an information that can be extracted from a tensor in OrtValue.
+    /// 
+    /// </summary>
+    public struct OrtTensorTypeAndShapeInfo
+    {
+        internal OrtTensorTypeAndShapeInfo(IntPtr handle)
+        {
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetTensorElementType(handle, out IntPtr elementType));
+            ElementDataType = (TensorElementType)elementType;
+
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetTensorShapeElementCount(handle, out UIntPtr count));
+            ElementCount = (long)count;
+
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetDimensionsCount(handle, out UIntPtr dimCount));
+
+            Shape = new long[(uint)dimCount];
+            NativeApiStatus.VerifySuccess(NativeMethods.OrtGetDimensions(handle, Shape, dimCount));
+        }
+
+        /// <summary>
+        /// Fetches tensor element data type
+        /// </summary>
+        /// <value>enum value for the data type</value>
+        public TensorElementType ElementDataType { get; private set; }
+
+        /// <summary>
+        /// Returns true if this data element is a string
+        /// </summary>
+        public bool IsString
+        {
+            get
+            {
+                return ElementDataType == TensorElementType.String;
+            }
+        }
+
+        /// <summary>
+        /// Fetches tensor element count based on the shape information.
+        /// </summary>
+        /// <returns>number of typed tensor elements</returns>
+        public long ElementCount { get; private set; }
+
+        /// <summary>
+        /// Fetches shape dimension count (rank) for the tensor.
+        /// </summary>
+        /// <returns>dim count</returns>
+        public int DimensionsCount { get { return Shape.Length; } }
+
+        /// <summary>
+        /// Tensor integer dimensions. Symbolic dimensions are represented as -1.
+        /// Use SymbolicDimensions to obtain symbolic dimensions.
```

Comment:
nit: comment needs update if this is only used for concrete Tensors with no symbolic dims

Reply:
Is this relevant if it's only used for concrete Tensors? `Symbolic dimensions are represented as -1.`